Add data defined override for 3d point symbol scale, rotation and translation

nyalldawson opened the pull request:
Adds data defined control over 3d point symbol x/y/z scaling, allowing for per-feature sizing of objects.

nyalldawson commented:
@wonder-sk While you're looking over PRs, mind taking a look and confirming whether this is heading in the right direction? If so, I'll add tests + cleanups

wonder-sk commented:
@nyalldawson the code looks sensible to me. One thing that comes to mind is that in the vertex shader, we should be also updating normal vectors of vertices when using data-defined scaling. We may live even without that, but the shading may be a little bit off for some shapes (when the scale factor is not the same for all x/y/z axes).

nyalldawson committed:
[feature] Data defined 3d point symbol scale, rotation and translation

wonder-sk reviewed (commented):
Great stuff! 👏 

Could you please add a rendering test as well?

Just to make sure if I understand the low-level changes...

The transformation of a vertex goes like this if we think of them as matrix multiplications:
```
V_final = T_geo * T_i * R_i * S_i * Z * V_input
```

where
- V_input: input coordinate of the instanced mesh
- Z: y-up to z-up flip
- T_i, R_i, S_i: per-instance translation, rotation and scale
- T_geo: translation within a chunk

Before this PR, the implementation was that:
- `pos` was T_geo
- `inst` was T_i * R_i * S_i * Z

And now it is:
- `pos` is T_geo * T_i
- `inst` is Z
- `scale` is S_i and `rotation` is R_i

Is that correct? Maybe it would be worth adding a short comment about the intended chain of transforms to the instanced.vert, so one does not always need to parse the whole thing 🙂 

I would also suggest the following renames in the instanced.vert shader:
- `pos` -> `instanceTranslation`
- `scale` -> `instanceScale`
- `rotation` -> `instanceRotation`

It would make the code a bit easier to digest (hopefully)

nyalldawson committed:
Rename instance shader variables for clarity

nyalldawson committed:
Rename more variables for clarity, tweak comments

nyalldawson committed:
When qt dumps shader code for error reporting, add line numbers

Makes it easier to debug problematic shader code, as the compiled
shader error line numbers won't correspond to the source code versions
of the shader

nyalldawson committed:
Remove unnecessary constant parameters from shader

nyalldawson committed:
Correct handle symbol scale, rotation in 3d scene export

nyalldawson committed:
Cleanup instanced shader by using defines

wonder-sk commented on src/3d/shaders/instanced.vert:
```diff
@@ -11,24 +17,71 @@ uniform mat4 modelMatrix;
-uniform mat4 inst;  // transform of individual object instance
-uniform mat4 instNormal;  // should be mat3 but Qt3D only supports mat4...
+uniform vec4 symbolRotation;
+uniform vec4 symbolScale;
```
If there isn't a good reason for it, I would keep both `instanceScale` and `symbolScale` as vec3, it looks like we only use .xyz from it anyway

wonder-sk reviewed (approved):
Things are looking great, thanks for the updates, and a nice touch with output of line numbers from a failing shader! 🙂

nyalldawson committed:
Use vec3 for scale

nyalldawson committed:
Add some qt6 test masks

nyalldawson committed:
Add unit tests for instanced point transforms, data defined transforms

nyalldawson committed:
Fix incorrect test reference file

The reference obj was incorrect and had the planes vertically oriented,
and only passed because the original fix was broken.

nyalldawson committed:
Handle transforms for instanced points in 3d exporter

nyalldawson committed:
Make model point symbols respect data defined properties

nyalldawson committed:
Fix clang-tidy warning
